internal/consumer/dlq.go

```diff
@@ -36,97 +38,148 @@ type (
-	if err != nil {
-		d.logger.Error("error sending message to DLQ", zap.Error(err))
+func newBufferedDLQ(topic, cluster string, producer sarama.AsyncProducer, scope tally.Scope, logger *zap.Logger) *bufferedErrorTopic {
+	return &bufferedErrorTopic{
+		topic: topic,
+		cluster: cluster,
+		producer: producer,
+		stopC: make(chan struct{}),
+		doneC: make(chan struct{}),
+		scope: scope,
+		logger: logger,
+		lifecycle: util.NewRunLifecycle(fmt.Sprintf("dlqProducer-%s-%s", topic, cluster), logger),
 	}
-	return err
 }
 
-func (d *dlqImpl) isRetryable(err error) bool {
-	return true
+// NewDLQNoop returns a DLQ that drops everything on the floor
+// this is used only when DLQ is not configured for a consumer
+func NewDLQNoop() DLQ {
+	return &dlqNoop{}
 }
+func (d *dlqNoop) Add(m kafka.Message) error { return nil }
+func (d *dlqNoop) Stop()                     {}
+func (d *dlqNoop) Start() error              { return nil }
 
-func newDLQRetryPolicy() backoff.RetryPolicy {
-	policy := backoff.NewExponentialRetryPolicy(dlqRetryInitialInterval)
-	policy.SetMaximumInterval(dlqRetryMaxInterval)
-	policy.SetExpirationInterval(backoff.NoInterval)
-	return policy
+func (d *bufferedErrorTopic) Start() error {
+	return d.lifecycle.Start(func() error {
+		go d.asyncProducerResponseLoop()
+		return nil
+	})
 }
 
-func (d *dlqImpl) newSaramaMessage(m kafka.Message) (*sarama.ProducerMessage, error) {
-	key, err := d.encodeDLQMetadata(m)
-	if err != nil {
-		return nil, err
-	}
-	return &sarama.ProducerMessage{
-		Topic: d.topic,
-		Key:   sarama.ByteEncoder(key),
-		Value: sarama.ByteEncoder(m.Value()),
-	}, nil
+func (d *bufferedErrorTopic) Stop() {
+	d.lifecycle.Stop(func() {
+		close(d.stopC)
+		<-d.doneC
+	})
 }
 
-func (d *dlqImpl) encodeDLQMetadata(m kafka.Message) ([]byte, error) {
+func (d *bufferedErrorTopic) Add(m kafka.Message) error {
 	metadata := &DLQMetadata{
-		RetryCount:  0,
+		RetryCount:  m.RetryCount() + 1,
 		Data:        m.Key(),
 		Topic:       m.Topic(),
 		Partition:   m.Partition(),
 		Offset:      m.Offset(),
 		TimestampNs: m.Timestamp().UnixNano(),
 	}
-	return proto.Marshal(metadata)
+
+	key, err := proto.Marshal(metadata)
+	if err != nil {
+		return err
+	}
+	value := m.Value()
+	// TODO (gteo): Use a channel pool
+	responseC := make(chan error)
+
+	sm := d.newSaramaMessage(key, value, responseC)
+	d.producer.Input() <- sm
+
+	select {
+	case err := <- responseC: // block until response is received
+		return err
+	case <- d.stopC:
+		return errors.New("closing producer failed to send")
+	}
 }
 
-// NewDLQNoop returns a DLQ that drops everything on the floor
-// this is used only when DLQ is not configured for a consumer
-func NewDLQNoop() DLQ {
-	return &dlqNoop{}
+func (d *bufferedErrorTopic) asyncProducerResponseLoop() {
+	for {
+		select {
+		case msg := <- d.producer.Successes():
+			if msg == nil {
+				continue
+			}
+			responseC, ok := msg.Metadata.(chan error)
+			if !ok {
+				continue
```

Comment:
emit an error log here, this should never happen

Reply:
Done